### PR TITLE
Don't instantiate user at module level

### PR DESCRIPTION
**Pull request by sgomes**
`lib/user` is part of several circular dependencies, so instantiation at the top level can lead to hard-to-find initialisation order problems.

Since `lib/user` follows a singleton pattern anyway, there's no need for performance considerations, and we can just invoke the default export as often as needed.

This PR is a follow-up to #43022, where @jsnajdr proposed this janitorial task.

#### Changes proposed in this Pull Request

* Remove top-level instantiation of the default export from `lib/user` across Calypso
* Update documentation to use preferred idiom

#### Testing instructions

This one is a bit hard to provide instructions for, since it touches several parts of Calypso. The unit tests + the E2E tests should provide a reasonable measure of safety.

### CODE REVIEW

**Review by nsakaimbo — CHANGES_REQUESTED** [review]
@sgomes About an hour ago, I just updated to latest Calypso master (Automattic/wp-calypso@6fcf873), build and e2e tests were good: Automattic/wp-desktop/pull/901

Since then, desktop CI and build for Calypso PR branches seem to be building fine:

#43066

#43067

This seems to indicate @jsnajdr's fix in #43022 worked. However CI for this branch specifically is currently failing: ([link](https://app.circleci.com/pipelines/github/Automattic/wp-desktop/15500/workflows/bc5787b2-363f-4443-8f0d-7841ad0817f7/jobs/63931/steps))

Seems like the desktop is sensitive to these circular-dependency changes?

**Comment by sgomes** [comment]
Looks like this PR introduces a fair bit of breakage. Investigating.

**Comment by sgomes** [comment]
Looks like things are failing because `lib/user` gets replaced with a `noop` in the server build, and I didn't take this into account.

**Comment by sgomes** [comment]
@nsakaimbo Looks like the failure wasn't desktop specific. While unit tests ran flawlessly, the server portions were crashing. It should work across the board now, but I'll wait for the tests before removing the `do not merge` label.

**Comment by sgomes** [comment]
@nsakaimbo The `wp-desktop` build passed, and the live branch works correctly too, now. I'll remove the label.

**Comment by nsakaimbo** [comment]
Awesome - thank you, @sgomes!

**Review by nsakaimbo — APPROVED** [review]
Thanks for fixing, @sgomes. I think the change makes sense conceptually, and CI across the board is ✅. LGTM.

File: client/lib/desktop/index.js
Comment (jsnajdr) [inline client/lib/desktop/index.js:105]: The check for `instanceof Array` is obsolete and can be removed. Before  #36781, the `User.prototype.clear` was setting `this.data = []` (I don't know why), but today it sets it to `false`.

File: client/lib/abtest/index.js
Comment (jsnajdr) [reply client/lib/abtest/index.js]: The `data?.ID` optional chaining is also not strictly necessary. `false` is a `Boolean` object with prototype and properties, and they can be dereferenced. We rely on that at a few other places in the codebase.

File: client/lib/signup/step-actions/index.js
Comment (jsnajdr) [inline client/lib/signup/step-actions/index.js:205]: The `lodash/get` here is not needed, by virtue of `false` being an object. If you don't want to (ab)use that (which I understand 🙂), optional chaining would still be a more modern option.

File: client/lib/translator-jumpstart/index.js
Comment (jsnajdr) [inline client/lib/translator-jumpstart/index.js:190]: `user().get().ID`

File: client/lib/translator-jumpstart/index.js
Comment (jsnajdr) [inline client/lib/translator-jumpstart/index.js:325]: `user().get().localeSlug`

File: client/me/account/main.jsx
Comment (jsnajdr) [inline client/me/account/main.jsx:401]: In this component, we should be retrieving user info from Redux. It's already `connect`ed.

File: client/me/purchases/controller.jsx
Comment (jsnajdr) [inline client/me/purchases/controller.jsx:32]: This should also use Redux: call a selector on `context.store.getState()`.

File: client/my-sites/checkout/checkout-thank-you/google-apps-details.jsx
Comment (jsnajdr) [inline client/my-sites/checkout/checkout-thank-you/google-apps-details.jsx:36]: This component should also use Redux.

(and also the `useTranslate` hook instead of calling `i18n.translate`, but that's unrelated)

File: client/my-sites/checkout/checkout-thank-you/jetpack-plan-details.jsx
Comment (jsnajdr) [inline client/my-sites/checkout/checkout-thank-you/jetpack-plan-details.jsx:25]: Should use Redux (and `useTranslate`)

**Review by jsnajdr — APPROVED** [review]
Posted quite a few suggestions, either related to cleanup in #36781, or suggestions to use Redux instead of `lib/user`. None of them is a blocker though.

File: client/lib/abtest/index.js
Comment (sgomes) [reply client/lib/abtest/index.js]: I switched to `user.get`, but I fear that relying on `false` having properties (which I didn't know, that's cool!) might lead to developer confusion, so I stuck with the optional chaining.

File: client/lib/desktop/index.js
Comment (sgomes) [reply client/lib/desktop/index.js]: Done! Also switched from `data` to `get()`.

File: client/lib/signup/step-actions/index.js
Comment (sgomes) [reply client/lib/signup/step-actions/index.js]: Switched to optional chaining 🙂

File: client/lib/translator-jumpstart/index.js
Comment (sgomes) [reply client/lib/translator-jumpstart/index.js]: Done!

File: client/lib/translator-jumpstart/index.js
Comment (sgomes) [reply client/lib/translator-jumpstart/index.js]: Done!

File: client/lib/abtest/index.js
Comment (jsnajdr) [inline client/lib/abtest/index.js:66]: `user()` is always guaranteed to return a valid `User` object. No need to check it. It's only the `user().get()` return value, i.e., `this.data` of the `User` singleton instance, that can be `false` for logged-out session.

File: client/lib/abtest/index.js
Comment (jsnajdr) [inline client/lib/abtest/index.js:302]: No need to check if `user()` is truthy.

File: client/lib/abtest/index.js
Comment (jsnajdr) [inline client/lib/abtest/index.js:313]: No need to optionally chain `user().get()`. Only the `ID` property access should be optional.

File: client/lib/abtest/index.js
Comment (sgomes) [reply client/lib/abtest/index.js:66]: That's sadly not true. I had assumed that at first but it broke the server side. Unfortunately, the server build aliases `lib/user` to a no-op, which means that `user()` ends up returning `undefined`, hence the need for the check.

File: client/lib/abtest/index.js
Comment (sgomes) [reply client/lib/abtest/index.js:302]: See above.

File: client/lib/abtest/index.js
Comment (sgomes) [reply client/lib/abtest/index.js:313]: See above.

**Comment by sgomes** [comment]
Added some comments regarding `user()` sometimes returning `undefined`, due to some `webpack` module replacement on the server build.

**Comment by sgomes** [comment]
Working on switching to Redux where @jsnajdr pointed it out.

File: client/me/account/main.jsx
Comment (sgomes) [reply client/me/account/main.jsx]: Done!

File: client/me/purchases/controller.jsx
Comment (sgomes) [reply client/me/purchases/controller.jsx]: Done!

File: client/my-sites/checkout/checkout-thank-you/google-apps-details.jsx
Comment (sgomes) [reply client/my-sites/checkout/checkout-thank-you/google-apps-details.jsx]: Done!

File: client/my-sites/checkout/checkout-thank-you/jetpack-plan-details.jsx
Comment (sgomes) [reply client/my-sites/checkout/checkout-thank-you/jetpack-plan-details.jsx]: Done! (Well, the user part, at least)

**Comment by sgomes** [comment]
Addressed all comments. PTAL, @jsnajdr!

File: client/lib/abtest/index.js
Comment (jsnajdr) [reply client/lib/abtest/index.js:66]: 🤦‍♂️ It's very likely that the server alias for `lib/user` is obsolete. `user.get()` used to trigger `user.fetch()`, which would be unwanted on server, but that doesn't happen any more.

And if we remove the static module-level calls to `user()` in this PR, that removes another reason why server-side `lib/user` could be harmful.

File: client/my-sites/checkout/checkout-thank-you/jetpack-plan-details.jsx
Comment (jsnajdr) [reply client/my-sites/checkout/checkout-thank-you/jetpack-plan-details.jsx]: Just keep in mind that the Redux refactor can also be in a followup PR. This one is already big enough 🙂

File: client/lib/abtest/index.js
Comment (jsnajdr) [reply client/lib/abtest/index.js:66]: By the way, it seems you do the `user()` check only in `lib/abtest` and nowhere else. Is `lib/abtest` really the only module affected by the server-side alias?

Comment (sgomes) [reply client/lib/abtest/index.js:66]: There are other places, I'm afraid 😕 I restored the check to all the places where it previously existed, in this commit, because I didn't want to untangle which ones were really needed and which ones weren't: https://github.com/Automattic/wp-calypso/pull/43069/commits/ccf62e9a04236196d3a93fd3adb63b990fe784c8

it includes signup and initial state as well.

What do you think the right course of action is here? Remove the alias and the checks as part of this PR as well?

Comment (jsnajdr) [reply client/lib/abtest/index.js:66]: > What do you think the right course of action is here?

Remove the alias and the check, but it totally deserves a standalone PR. This one is already snowballing into areas you probably didn't plan to address in the beginning.

Comment (sgomes) [reply client/lib/abtest/index.js:66]: No problem! I'll prepare another PR with those changes, then 🙂 Thanks!

**Comment by sgomes** [comment]
After discussing with @jsnajdr, I think this PR is good to go, and I'll follow up with a new PR to tackle the `lib/user` aliasing on server.

I'll go ahead and merge this PR as-is after all tests are green (restarted the failing E2E tests, which appear to be displaying a transient failure).

**Comment by jsnajdr** [comment]
> I think this PR is good to go

:+1:

**Comment by sgomes** [comment]
After the third attempt, the E2E tests completed successfully. Merging. Thank you for the reviews, @nsakaimbo and @jsnajdr!  🙇

**Pull request merged**
